### PR TITLE
Updated the Android billing library to version 5

### LINKED ISSUE #51: From November 1, 2022, all app updates must use Billing Library version 4 or newer.
The extension must be updated to Google Billing Library version 4 (or 5) by November 1, 2022 or earlier.
Thanks!!!
image

### PR DESCRIPTION
**Pull request by britzl**
Fixes #51

### CODE REVIEW

File: extension-iap/src/iap_android.cpp [hunk 1]
```diff
@@ -68,36 +70,46 @@ static int IAP_List(lua_State* L)
...
+    DM_LUA_STACK_CHECK(L, 0);
 
+    int top = lua_gettop(L);
     const char* id = luaL_checkstring(L, 1);
+    const char* token = "";
+
+    if (top >= 2 && lua_istable(L, 2)) {
+        luaL_checktype(L, 2, LUA_TTABLE);
+        lua_pushvalue(L, 2);
+        lua_getfield(L, -1, "token");
```
Comment (AGulev) [inline extension-iap/src/iap_android.cpp:87]: What this token is? Should we add it in doc and script_api?

File: extension-iap/src/java/com/defold/iap/IapGooglePlay.java [hunk 2]
```diff
@@ -110,16 +117,54 @@ private String convertPurchase(Purchase purchase) {
...
+            List<PricingPhase> pricingPhases = details.getPricingPhases().getPricingPhaseList();
+            for (PricingPhase pricingPhase : pricingPhases) {
+                JSONObject o = new JSONObject();
+                o.put("price_string", pricingPhase.getFormattedPrice());
+                o.put("price", pricingPhase.getPriceAmountMicros() * 0.000001);
+                o.put("currency_code", pricingPhase.getPriceCurrencyCode());
+                o.put("billing_period", pricingPhase.getBillingPeriod());
+                o.put("billing_cycle_count", pricingPhase.getBillingCycleCount());
+                o.put("recurrence_mode", pricingPhase.getRecurrenceMode());
+                a.put(o);
```
Comment (AGulev) [inline extension-iap/src/java/com/defold/iap/IapGooglePlay.java:132]: Maybe we should add info in doc about new fields in response?

File: extension-iap/src/java/com/defold/iap/IapGooglePlay.java [hunk 3]
```diff
@@ -356,30 +414,43 @@ public void onSkuDetailsResponse(BillingResult billingResult, List<SkuDetails> s
...
-                    listener.onSkuDetailsResponse(billingResult, allSkuDetails);
+                    listener.onProductDetailsResponse(billingResult, allProductDetails);
                 }
             }
         };
-        billingClient.querySkuDetailsAsync(SkuDetailsParams.newBuilder().setSkusList(skuList).setType(SkuType.INAPP).build(), detailsListener);
-        billingClient.querySkuDetailsAsync(SkuDetailsParams.newBuilder().setSkusList(skuList).setType(SkuType.SUBS).build(), detailsListener);
+
+        // we don't know if a product is a subscription or inapp product type
+        // instread we create two product lists from the same set of products and use INAPP for one and SUBS for the other
```
Comment (AGulev) [inline extension-iap/src/java/com/defold/iap/IapGooglePlay.java:441]: hmm, is it valid way to do it? Shouldn't we ask user to specify it?

File: main/main.gui_script [hunk 4]
```diff
@@ -66,6 +66,7 @@ local function list()
 		for k,p in pairs(products) do
 			available_items[p.ident] = p
 			log("Item %s", p.ident)
+			pprint(p)
 			local button = item_buttons[p.ident]
 			if button then
```
Comment (AGulev) [inline main/main.gui_script:71]: should we add an example for buying with token?

**Review by AGulev — APPROVED** [review]
LGTM.
I have a few questions

File: extension-iap/src/iap_android.cpp (hunk 1 above)
Comment (britzl) [reply extension-iap/src/iap_android.cpp:87]: Added. It's a subscription specific thing to distinguish between different offers that the user is eligible for.

https://developer.android.com/reference/com/android/billingclient/api/BillingFlowParams.ProductDetailsParams.Builder#setOfferToken(java.lang.String)

File: extension-iap/src/java/com/defold/iap/IapGooglePlay.java (hunk 3 above)
Comment (britzl) [reply extension-iap/src/java/com/defold/iap/IapGooglePlay.java:455]: This is the way we did it in previous versions of the extension as well.

File: extension-iap/src/java/com/defold/iap/IapGooglePlay.java (hunk 2 above)
Comment (britzl) [reply extension-iap/src/java/com/defold/iap/IapGooglePlay.java:144]: Added!

**Pull request merged**